### PR TITLE
Graphite: Migrate to React (part 1: move state to redux)

### PR DESCRIPTION
**Pull request by ifrost**
This is first PR for Graphite to React migration. I'm trying to replace the engine while driving the car here. More details in #36481.

First step is to move all logic from Angular controller to redux state/reduces. They will be used by React components later. No new components, no changes to the UI or the behavior. The logic is almost identical - I just moved functions to different places. I kept empty functions in GraphiteQueryCtrl with some comments for easier diffing.

All functions from GraphiteQueryCtrl have been moved to separate files:
- state/actions - with all available UI actions
- state/providers - hooks providing data for autocomplete
- state/store - main store and reducers
- state/helpers - common logic used by reducers and providers

How to test it?

Graphite is a complex beast. You can use [manual test scenarios ](https://gist.github.com/ifrost/e8faa2976caf610e9e03ec815ce83d82) or simply play with it compering the behavior with main/play.

### CODE REVIEW

**Comment by torkelo** [comment]
Nice! 
did you take a look at the Elastic query editor? think it uses useReducer instead of redux ,  written by @Elfo404 . maybe there are some patterns there we can reuse. not that there is much different with functional components now with useReducer  vs useDispatch

**Comment by ifrost** [comment]
> Nice!
> did you take a look at the Elastic query editor? think it uses useReducer instead of redux , written by @Elfo404 . maybe there are some patterns there we can reuse. not that there is much different with functional components now with useReducer vs useDispatch

Yes, I think it makes total sense to make it very similar to Elastic and maybe share some code for creating ContextProviders. I'll try to tackle it when I get to the point when I can remove Angular controller. As an intermediate step I thought I could simply pass down `state` / `dispatch` to make it easier to refactor all component bottom-up and add the root React component at the end.

File: public/app/plugins/datasource/graphite/state/providers.ts [hunk 1]
```diff
@@ -0,0 +1,190 @@
...
+  });
+
+  return altSegments.concat(...tagSegments);
+}
+
+function removeTaggedEntry(altSegments: any[]) {
+  remove(altSegments, (s) => s.value === '_tagged');
+}
+
+function mapToDropdownOptions(results: any[]) {
```
Comment (Elfo404) [inline public/app/plugins/datasource/graphite/state/providers.ts:186]:
```suggestion
function mapToDropdownOptions(results: string[]) {
```

File: public/app/plugins/datasource/graphite/state/providers.ts [hunk 2]
```diff
@@ -0,0 +1,190 @@
...
+ * It may be:
+ * - mixed list of metrics and tags (only when nothing was selected)
+ * - list of metric names (if a metric name was selected for this segment)
+ */
+export async function getAltSegments(
+  state: GraphiteQueryEditorState,
+  index: number,
+  prefix: string
+): Promise<GraphiteSegment[]> {
+  let query = prefix && prefix.length > 0 ? '*' + prefix + '*' : '*';
```
[the remark this answers is not in the document]
Comment (ifrost) [reply public/app/plugins/datasource/graphite/state/providers.ts:27]: Prefix might be an empty string but should always be defined. I'll pick the second suggestion 👍

**Comment by ifrost** [comment]
@Elfo404 - I've addressed your suggestions 👍 I've also added more types. It's surfaced a bug with missing `target` property when checking when to refresh the panel [here](https://github.com/grafana/grafana/pull/36484/commits/a952d47c8abe08942ae0b388e959384ff1e244a9#diff-403fc986757091098706d2088928dae68ebe907e5fc50b0b0b54d3f5fdf1aa9aR183) 😎

**Review by aocenas — COMMENTED** [review]
Nice, I did not test it thoroughly but like where this is going. Also like explanatory comments in the code makes following the changes much easier.

**Comment by ifrost** [comment]
a3f13ee: small change to fix loading function definitions. They should be loaded before the model is created to correctly recognize already defined functions. 

query_ctrl.tsx (main)
```typescript
this.datasource.waitForFuncDefsLoaded().then(() => {                              // load functions
  this.queryModel = new GraphiteQuery(this.datasource, this.target, templateSrv); // create model
  this.buildSegments(false);                                                      // build ui
});
```

store.tsx (branch)
```typescript
  if (actions.init.match(action)) {
    await deps.datasource.waitForFuncDefsLoaded();                                    // load functions
    state = {
      ...
      queryModel: new GraphiteQuery(deps.datasource, deps.target, getTemplateSrv()),  // create model
    };
    await buildSegments(state, false);                                                // build ui
  }
```

**Review by Elfo404 — APPROVED** [review]
Just a small nit (aside from the async things discussed offline), but looks great and seems to work as expected! 🎉 

I really like how you migrated the state before during UI work, great stuff in here! 🔥

**Review by aocenas — APPROVED** [review]
Tried it a bit and seems to work the same as previous version.

There is a bug that seems to be in both versions where adding a tag and removing it then prevents adding a metric:

https://user-images.githubusercontent.com/1014802/126306253-f6594f47-de51-4ba7-80ca-871e51efe3b6.mp4

I guess we can later fix it separatelly

File: public/app/plugins/datasource/graphite/state/store.ts [hunk 3]
```diff
@@ -0,0 +1,177 @@
...
+  error: Error | null;
+
+  tagsAutoCompleteErrorShown: boolean;
+  metricAutoCompleteErrorShown: boolean;
+};
+
+type Action = {
+  type: string;
+  payload: any;
+};
```
[the remark this answers is not in the document]
Comment (ifrost) [reply public/app/plugins/datasource/graphite/state/store.ts:52]: > Tried it a bit and seems to work the same as previous version.
> 
> There is a bug that seems to be in both versions where adding a tag and removing it then prevents adding a metric:
> 
>  Peek.2021-07-20.12-07.mp4 
> I guess we can later fix it separatelly

Yep - it's the bug in the current implementation. I've added a note to test cases scenarios. I'll try to fix different bugs once all components are migrated.

**Comment by ifrost** [comment]
> Tried it a bit and seems to work the same as previous version.
> 
> There is a bug that seems to be in both versions where adding a tag and removing it then prevents adding a metric:
> 
>  Peek.2021-07-20.12-07.mp4 
> I guess we can later fix it separatelly

Yep - it's the bug in the current implementation. I've added a note to test cases scenarios. I'll try to fix different bugs once all components are migrated.

File: public/app/plugins/datasource/graphite/state/store.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (ifrost) [reply public/app/plugins/datasource/graphite/state/store.ts]: Absolutely - update with feed0cd

File: public/app/plugins/datasource/graphite/state/providers.ts [hunk 4]
```diff
@@ -0,0 +1,190 @@
...
+    tagsAsSegments = [];
+    await handleTagsAutoCompleteError(state, err);
+  }
+
+  return tagsAsSegments;
+}
+
+export async function getTagValues(
+  state: GraphiteQueryEditorState,
+  tag: { key: any },
```
[the remark this answers is not in the document]
Comment (ifrost) [reply public/app/plugins/datasource/graphite/state/providers.ts]: Good spot - c18ff18

**Pull request merged**